Add Chia BLS

ChihChengLiang opened the pull request:
### What was wrong?

Add Chia BLS

### How was it fixed?

Note that some API change

- Add validate and validate_multiple to distinguish from verify and verify_multiple.


#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](https://preview.redd.it/d702kl24pe931.jpg?width=640&crop=smart&auto=webp&s=efd8ca6917d3a4ba71b7780008347275fee800f5)

mhchia commented on tests/eth2/conftest.py:
```diff
@@ -117,7 +120,7 @@ def _get_privkey_for(self, index):
            If using ``i``, the combinations of privkeys would not lead to unique pubkeys.
         """
-        privkey = 2**index
+        privkey = int.from_bytes(hash_eth2(index.to_bytes(32, 'little')), 'little') % 2**32
```
Is this used to ensure 1) the generated private key is less than `2**32` and 2) the combinations do not easily collide to the other generated private keys?

ChihChengLiang replied:
Yes

mhchia commented on tests/eth2/core/beacon/tools/builder/test_validator.py:
```diff
@@ -75,4 +75,10 @@ def test_aggregate_votes(votes_count, random, privkeys, pubkeys):
+    if votes_count != 0:
+        bls.validate(message_hash, aggregated_pubs, sigs, domain)
+    else:
+        # EMPTY_SIGNATURE is consider invalid
```
nit: 
```suggestion
        # EMPTY_SIGNATURE is considered invalid
```

ralexstokes commented on eth2/_utils/bls.py:
```diff
@@ -0,0 +1,126 @@
+
+
+class BaseBLSBackend(ABC):
+
```
nitpick: remove this newline?

ralexstokes commented on eth2/_utils/bls.py:
```diff
@@ -0,0 +1,126 @@
+
```
nitpick: remove newline?

ralexstokes commented on eth2/_utils/bls_bindings/chia_network/api.py:
```diff
@@ -10,13 +10,50 @@
 
-def _privkey_int_to_bytes(privkey: int) -> bytes:
-    return privkey.to_bytes(bls_chia.PrivateKey.PRIVATE_KEY_SIZE, "big")
+def _privkey_from_int(privkey: int) -> 'bls_chia.PrivateKey':
```
i tried removing the `'` so that we are directly referencing the class and it seems to work

can we update the typing so that we just use the symbol directly?

ralexstokes commented on tests/eth2/conftest.py:
```diff
@@ -117,7 +120,7 @@ def _get_privkey_for(self, index):
            If using ``i``, the combinations of privkeys would not lead to unique pubkeys.
         """
-        privkey = 2**index
+        privkey = int.from_bytes(hash_eth2(index.to_bytes(32, 'little')), 'little') % 2**32
```
+1 on asking about rationale... if we do keep this change then we need to update the comment about how keys are selected/generated (you can just delete it if its out of date)

also -- i think we can drop the `% 2 ** 32` here right? we never have a number greater or equal to `2**32` here

ChihChengLiang replied:
Nice catch

ChihChengLiang replied:
Ah, so the purpose here is to generate a valid private key. It should be greater than 0 but less than the curve order, which is a number between `2**254` and `2**255`. 
(It's 32 "bytes" so 2 ** 32 is so wrong 🙈)

Without the modulus, it's really easy to get invalid private key that is greater than the curve order.
```python
>>> from py_ecc.optimized_bls12_381 import curve_order
>>> from eth2._utils.hash import hash_eth2
>>> len([index for index in range(1000) if int.from_bytes(hash_eth2(index.to_bytes(32, 'little')), 'little')> curve_order])
512
```

ralexstokes commented on tests/eth2/fixtures-tests/state-fixtures/test_minimal_state.py:
```diff
@@ -59,8 +59,8 @@ def mock_bls(mocker, request):
 
-    mocker.patch('py_ecc.bls.verify', side_effect=mock_bls_verify)
-    mocker.patch('py_ecc.bls.verify_multiple', side_effect=mock_bls_verify_multiple)
+    mocker.patch('eth2._utils.bls.eth2_bls.verify', side_effect=mock_bls_verify)
```
see other comment about use backends here, rather than mocking

ralexstokes commented on tests/eth2/integration/test_demo.py:
```diff
@@ -45,8 +45,8 @@ def mock_bls_verify_multiple(pubkeys,
 def mock_bls(mocker, request):
-    mocker.patch('py_ecc.bls.verify', side_effect=mock_bls_verify)
-    mocker.patch('py_ecc.bls.verify_multiple', side_effect=mock_bls_verify_multiple)
+    mocker.patch('eth2._utils.bls.eth2_bls.verify', side_effect=mock_bls_verify)
```
the idea w/ moving to this `BLSBackend` pattern is so that we don't have to mock out bls -- we can just select the BLS backend we want local to this module...

i would think we would rather prefer an import at the top of this file that says something like:

`from eth2._utils.bls import NoOpVerifyBackend as bls`

or something like this

ChihChengLiang commented:
@ralexstokes Answer a part of your comments

> 1. do we want to include a BLSBackend for the py_ecc impl? it is a little hard to compare the chia 
backend w/ the py_ecc backend as the py_ecc backend disappears in this PR!

Originally not planing to add them now to speed up the adoption of Chia bls. But happy to add them if that helps the review.

> 2. but i'm wondering if all of the validation code we do in api.py is required?

The bls binding prints only runtime error, which is really hard to debug. Adding the validations helps rule out the known reasons. They are not necessarily to stay in api.py though.

> 3. dropping the bls_chia prefix in api.py.

Done 👍

> 4. what is the rationale for the validate on top of the verify?

The idea is to move toward the pattern described in #296. I like the simplicity of returning bool too. But the `validate` API shows the public key, signature, message, and domain when raising a validation error. In `verify` API those information need to be rewritten again in the catch section of the caller.

ralexstokes commented:
> @ralexstokes Answer a part of your comments
> 
> > 1. do we want to include a BLSBackend for the py_ecc impl? it is a little hard to compare the chia
> >    backend w/ the py_ecc backend as the py_ecc backend disappears in this PR!
> 
> Originally not planing to add them now to speed up the adoption of Chia bls. But happy to add them if that helps the review.

it isn't urgent but we will almost certainly want it -- i doubt we will have no further use for `py_ecc` ever. we don't have to add it as a backend in this PR but something to keep an eye on

> 
> > 1. but i'm wondering if all of the validation code we do in api.py is required?
> 
> The bls binding prints only runtime error, which is really hard to debug. Adding the validations helps rule out the known reasons. They are not necessarily to stay in api.py though.

cool makes sense to keep it then

> 
> > 1. dropping the bls_chia prefix in api.py.
> 
> Done 👍
> 
> > 1. what is the rationale for the validate on top of the verify?
> 
> The idea is to move toward the pattern described in #296. I like the simplicity of returning bool too. But the `validate` API shows the public key, signature, message, and domain when raising a validation error. In `verify` API those information need to be rewritten again in the catch section of the caller.

gotcha! thanks for explaining

ChihChengLiang commented:
Hi @ralexstokes @mhchia, this PR is ready to be rereviewed again. Some major changes:

- The directory structure has changed a lot now.
- Added PyECCBackend
- Removed all bls mockers.
- Removed bls-binding test. Replaced it with a test that run through all available backends.

ralexstokes commented on eth2/_utils/bls/__init__.py:
```diff
@@ -0,0 +1,115 @@
+
+    @classmethod
+    def use_default_backend(cls) -> None:
+        cls.backend = DEFAULT_BACKEND
```
minor but what do you think about:

```python
def use_default_backend(cls) -> None:
    cls.use(DEFAULT_BACKEND)
```
could be that in the future we have more stuff to do then just set one class var

ralexstokes commented on eth2/_utils/bls/__init__.py:
```diff
@@ -0,0 +1,115 @@
+
+    @classmethod
+    def use_noop_backend(cls) -> None:
+        cls.backend = NoOpBackend
```
same about unifying setter logic here

ralexstokes commented on eth2/_utils/bls/__init__.py:
```diff
@@ -0,0 +1,115 @@
+
+
+eth2_bls = Eth2BLS
+eth2_bls.use_default_backend()
```
why not just:

```python
eth2_bls = Eth2BLS() # selects default backend for you
```
```

ralexstokes commented on eth2/_utils/bls/__init__.py:
```diff
@@ -0,0 +1,115 @@
+            )
+
+
+eth2_bls = Eth2BLS
```
is it wrong to just call this variable `bls`?

ralexstokes replied:
(it would simplify imports for callers)

ChihChengLiang commented:
Thank you all for the review
